apm-agent-core/src/test/java/specs/ApiKeysStepsDefinitions.java

```diff
@@ -0,0 +1,160 @@
+package specs;
+
+import co.elastic.apm.agent.configuration.converter.TimeDuration;
+import co.elastic.apm.agent.report.ApmServerClient;
+import co.elastic.apm.agent.report.HttpUtils;
+import co.elastic.apm.agent.report.ReporterConfiguration;
+import com.github.tomakehurst.wiremock.WireMockServer;
+import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
+import com.github.tomakehurst.wiremock.extension.responsetemplating.ResponseTemplateTransformer;
+import io.cucumber.java.After;
+import io.cucumber.java.Before;
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Given;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
+import org.stagemonitor.configuration.ConfigurationOption;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collections;
+
+import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
+import static com.github.tomakehurst.wiremock.client.WireMock.get;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Fail.fail;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ApiKeysStepsDefinitions {
+
+    // those constants should be provided by the gherkin test, not using hard-coded values
+    // as it would require to update the test definition, we'll keep it as-is for now.
+    private static final String SECRET_TOKEN = "secr3tT0ken";
+    private static final String API_KEY = "@p1Key";
+
+    // so far, only reporter and it's configuration is being tested
+    private ReporterConfiguration configuration = null;
+
+    private WireMockServer server = new WireMockServer(WireMockConfiguration.options()
+        .extensions(new ResponseTemplateTransformer(false))
+        .dynamicPort());
+
+    @Before
+    public void init() {
+        server.stubFor(get(urlEqualTo("/"))
+            .willReturn(aResponse()
+                .withTransformers("response-template")
+                // just send back auth header (if any) for easy parsing on client side
+                .withBody("{{request.headers.Authorization}}")));
+
+        server.start();
+    }
+
+    @After
+    public void cleanup() {
+        server.stop();
+    }
+
+    // Init
+
+    @Given("an agent")
+    public void initAgent() {
+        // we just initialize configuration as reporter is initialized lazily
+        configuration = mock(ReporterConfiguration.class);
```

Comment:
works fine as is but you could alternatively do `SpyConfiguration.createSpyConfig().getConfig(ReporterConfiguration.class);`.

This lets you avoid having to re-specify the default values (the `just required to avoid NPEs` step)

Reply:
Nice, I never took time to investigate differences between spies and mocks in mockito, now it's more clear :-)